int_scan.cpp

```diff
@@ -0,0 +1,115 @@
+#include "int_scan.hpp"
+#include <cassert>
+#include <ctype.h>
+
+static const int OCT = 8;
+static const int DEC = 10;
+static const int HEX = 16;
+
+/*
+The function returns a member of the Base enumeration, which represents the base
+of the input string. If the input string starts with the character 0 (zero),
+then the input string represents an octal number. If the input string starts
+with the characters 0x or 0X (zero x/X), then the input string represents a
+hexadecimal number. If the input string does not have one of these prefixes,
+then the input string represents a decimal number. The function starts looking
+for the prefix from the index specified by the position pointer. The function
+also returns the position of the next character of the input to be read after
+the prefix (if it exists).
+*/
+static int check_base(const char *input, int *position) {
+  int pos = 0;
+  if (input[pos] != '0') {
+    *position = pos;
+    return DEC;
+  }
+  pos++;
+  if (tolower(input[pos]) == 'x') {
+    pos++;
+    *position = pos;
+    return HEX;
+  }
+  *position = pos;
+  return OCT;
+}
+
+/*
+The function returns the integer value of the digit character passed as an
+argument. The digit's base is 10. If the digit is out of bounds, the function
+returns -1.
+*/
+static int base10_char_to_dec(char digit) {
+  if ('0' <= digit && digit <= '9')
+    return digit - '0';
+  return -1;
+}
+
+/*
+The function returns the integer value of the digit character passed as an
+argument. The digit's base is 8. If the digit is out of bounds, the function
+returns -1.
+*/
+static int base8_char_to_dec(char digit) {
+  if ('0' <= digit && digit <= '7')
+    return digit - '0';
+  return -1;
+}
+
+/*
+The function returns the integer value of the digit character passed as an
+argument. The digit's base is 16. If the digit is out of bounds, the function
+returns -1. Both uppercase and lowercase letters a-f and A-F are considered to
+be valid.
+*/
+static int base16_char_to_dec(char digit) {
+  if ('0' <= digit && digit <= '9')
+    return digit - '0';
+  char ldigit = tolower(digit);
+  if ('a' <= ldigit && ldigit <= 'f')
+    return 10 + ldigit - 'a';
+  return -1;
+}
+
+/*
+The function returns the integer value of the digit character passed as an
+argument depending on the digit's base. The digit's base is specified by the
+num_base argument of the function If the digit is out of bounds, the function
+returns -1.
+*/
+static int char_to_dec(char digit, const int num_base) {
+  if (num_base == DEC)
+    return base10_char_to_dec(digit);
+  if (num_base == OCT)
+    return base8_char_to_dec(digit);
+  assert(num_base == HEX);
+  return base16_char_to_dec(digit);
+}
```

Comment:
This function could use a map that maps digits to their values. Something like:
```cpp
uint8_t char_to_digit[256] = {
    ['0'] = 0,
    ['1'] = 1,
    ['2'] = 2,
    ['3'] = 3,
    ['4'] = 4,
    ['5'] = 5,
    ['6'] = 6,
    ['7'] = 7,
    ['8'] = 8,
    ['9'] = 9,
    ['a'] = 10, ['A'] = 10,
    ['b'] = 11, ['B'] = 11,
    ['c'] = 12, ['C'] = 12,
    ['d'] = 13, ['D'] = 13,
    ['e'] = 14, ['E'] = 14,
    ['f'] = 15, ['F'] = 15,
};
```
Then, you can simply do:
```cpp
int val = char_to_digit[digit];
if ((val == 0 && digit != 0) || val >= num_base)
  return -1;
return val;
```

Reply:
That was a cool idea. First time i saw designated initialization syntax so i looked it up. It turns out that designated initialization for arrays is not valid for c++ as opposed to c.
`https://en.cppreference.com/w/cpp/language/aggregate_initialization`

Reply:
You're right. Well, they're supported in C++20 but we don't want to be that progressive..
In the `main()` you're testing that, just initialize the array. This function is essentially going to be part of the lexer and the lexer will need a general initialization function anyway. So, we will throw this initialization there.

_Note_: You can use the syntax: `[link text](link url)` to post links. Alternatively, you can highlight the the text you want to act as link and use the icon near the `<>` above.

Reply:
Should the array be declared globally or locally inside the main function and then passed as an argument to the int_scan function?

Reply:
Globally, because the `main` function will act as a "filler" for the array.